themes/default/content/learn/pulumi-fundamentals/create-docker-images/index.md

```diff
@@ -0,0 +1,244 @@
+contain a `Dockerfile` that builds the application images.
+
+Let's examine the backend `Dockerfile` in `app/backend/Dockerfile`:
+
+```docker
+FROM node:14
+
+# Create app directory
+WORKDIR /usr/src/app
+
+COPY ./src/package*.json ./
+RUN npm install
+COPY ./src .
+RUN npm build
+EXPOSE 3000
+
+CMD [ "npm", "start" ]
+```
+
+This `Dockerfile` copies the REST backend into the Docker filesystem, installs
+the dependencies, and builds the image. Note that port 3000 must be open on your
+host machine.
+
+## Build your Docker Image with Pulumi
+
+Before we start writing our Pulumi program, we need to install the right
+provider. In this case, we want to use the `pulumi_docker` provider for Python,
+our language host. It's always good practice for Python to work inside a virtual
+environment, or venv, so let's activate our venv and use `pip` to install the
+provider along with the main Pulumi package:
+
+```bash
+cd ../
+source venv/bin/activate
+pip3 install pulumi_docker
+```
+
+You should see output showing the provider package being installed, just like
+for any Python package install. You'll want to ensure the package gets added to
+the `requirements.txt` file at some point.
+
+Back inside your Pulumi program, let's build your first Docker image. Remember
+that a Pulumi program is the code that defines the desired state of your
+infrastructure using a general-purpose programming language. In this case, we're
+using Python, so our main file is {{% langfile %}}. Inside your program's
+{{% langfile %}} file, use any editor to add the following code:
+
+{{< chooser language "python" / >}}
+
+{{% choosable language python %}}
+
+```python
+import os
+import pulumi
+import pulumi_docker as docker
+
+stack = pulumi.get_stack()
+
+# build our backend image!
+backend_image_name = "backend"
+backend = docker.Image("backend",
+                        build=docker.DockerBuild(context=f"{os.getcwd()}/app/backend"),
+                        image_name=f"{backend_image_name}:{stack}",
+                        skip_push=True
+                        )
+```
+{{% /choosable %}}
+
+In this file, we import the main Pulumi package and the Docker provider. Then,
+we figure out which stack we're operating against, and populate the `stack`
+variable for later use. When we build our backend image, we name it in our stack
+as "backend" before passing some arguments to the Docker provider. The Docker
+provider uses the build context and the image name to build an image, and it
+does not push the image up anywhere.
+
+Notice that we're mixing in some language constructs in here like `os.getcwd()`.
+With Pulumi, we have access to the full language ecosystem, including
+third-party libraries. Pulumi also has typing support, so you can use the tools
+in your favorite IDE, like completion, to verify that you're using the correct
+types for any inputs you're using. Pretty cool!
+
+Run `pulumi up`, and Pulumi should build your Docker image. First, though, it
```

Comment:
When I run these `pulumi up`s myself, it takes _many_ seconds before I see any output from the CLI -- presumably because of what Docker's doing in the background. It's just long enough to make me worry the CLI might actually be stuck. Eventually, everything completes, so I'm not sure it's worth calling out on the page necessarily, but since I noticed it a couple of times during the lesson, it seemed worth mentioning.

Reply:
Good callouts. I'll give it its own line and add a callout note.